search/util.go

```diff
@@ -67,3 +69,32 @@ func MergeFieldTermLocations(dest []FieldTermLocation, matches []*DocumentMatch)
 
 	return dest
 }
+
+const SearchIOStatsCallbackKey = "_search_io_stats_callback_key"
+
+type SearchIOStatsCallbackFunc func(uint64)
+
+// The callback signature is (message, queryType, cost) which allows
+// the caller to act on a particular query type and what its the associated
+// cost of an operation. "add" indicates to increment the cost for the query
```

Comment:
Values of message argument (like "add" and "done" ) aren't defined by bleve, it is upto the user of bleve.

If you have specified "add" and "done" as examples of message argument value, then we can frame it that way

Reply:
hmm, the cost callback according to me includes very basic set of messages, which are add, done and maybe abort, since its mainly enabling the caller to track the costs incrementally and act upon that along with a "done" type of a message that will help the caller to act upon the tracked bytes until now. perhaps an abort invocation can be done when there is a ctx.Done() in the collector phase (which would happen when the search's context is cancelled i believe) in that case, the caller can handle separately perhaps.

Reply:
Got your point. 
Bleve will invoke the callback with these messages "add", "done", "abort" etc ... 
And it is the responsibility of the author of the callback to properly handle these messages.

In this case, we can frame our comment like this:
"Implementation of SearchIncrementalCostCallbackFn should handle the following messages
    - add: "description ..."
    - abort:  description ...
    - done:  description ...
"

We can even assign IDs to these messages and export from bleve

type IncCostMessage uint

const (
	AddM = Service(1 << iota) // description
	AbortM // description
	DoneM // description
)

type SearchIncrementalCostCallbackFn func(IncCostMessage, string, uint64)

Thoughts ?

Reply:
yeah i've renamed the messages and query types, thanks